File: src/damage_type.py
WEAK_DAMAGE, DEFAULT_DAMAGE, STRONG_DAMAGE = 1, 2, 3

class DamageType(object):
    """
    Defines a damage type, which determines how much damage an attack does against a specified enemy.
    """
    def __init__(self, name: str, strong_against: str):
        self.name: str = name
        self.strong_against: callable = lambda dt : dt.name == strong_against
    
    def damage_against(self, other) -> int:
        if other is None: return DEFAULT_DAMAGE
        if isinstance(other, str): return self.damage_against(DAMAGE_TYPES[other])
        if self.strong_against(other):
            return STRONG_DAMAGE
        if other.strong_against(self):
            return WEAK_DAMAGE
        return DEFAULT_DAMAGE
    
    def __str__(self):
        return "DamageType(" + self.name + ")"
    
    def __repr__(self):
        return self.__str__()
    
_types: dict[DamageType, DamageType] = {
    "None": "not strong against anything",
    "Fire": "Frost",
    "Water": "Fire",
    "Earth": "Water",
    "Wind": "Earth",
    "Lightning": "Wind",
    "Frost": "Lightning"
}

DAMAGE_TYPES: dict[str, DamageType] = {
    k: DamageType(k, v) for k, v in _types.items()
}

def damage_for(attackerType: str, defenderType: str):
    if attackerType not in DAMAGE_TYPES or defenderType not in DAMAGE_TYPES:
        return DEFAULT_DAMAGE
    return DAMAGE_TYPES[attackerType].damage_against(defenderType)

File: src/DamageType.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace epigenetic_agency;
public enum Damage
{
    Weak = 1,
    Default = 2,
    Strong = 3
}
public enum DamageType
{
    None,
    Fire,
    Water,
    Earth,
    Wind,
    Lightning,
    Frost
}
public static class DamageTypes
{
    public static IEnumerable<DamageType> All
        => Enum.GetValues<DamageType>();
    public static readonly Dictionary<DamageType, DamageType?> TypeStrengths = new()
    {
        { DamageType.None, null },
        { DamageType.Fire,      DamageType.Frost    },
        { DamageType.Water,     DamageType.Fire     },
        { DamageType.Earth,     DamageType.Water    },
        { DamageType.Wind,      DamageType.Earth    },
        { DamageType.Lightning, DamageType.Wind     },
        { DamageType.Frost,     DamageType.Fire     }
    };
    public static bool IsStrongAgainst(this DamageType attacker, DamageType defender)
        => TypeStrengths[attacker] == defender;
    public static int DamageAgainst(this DamageType attacker, DamageType defender)
    {
        if (attacker.IsStrongAgainst(defender))
            return (int)Damage.Strong;
        if (defender.IsStrongAgainst(attacker))
            return (int)Damage.Weak;
        return (int)Damage.Default;
    }
}
